### ISSUE TITLE
Single line if statement going to multiple lines should keep the header on one line if able

### ISSUE THREAD

**Issue by dsherret** [dprint/dprint-plugin-typescript#46]
**Describe the bug**

Deno version: 1.3.1

**Input Code**

```ts
if (!botCache.helpers.isAdmin(message, targetGuildSettings)) return addReaction(message.channelID, message.id, "❌");
```

**Expected Output**

```ts
if (!botCache.helpers.isAdmin(message, targetGuildSettings)) {
    return addReaction(message.channelID, message.id, "❌");
}
```

**Actual Output**

```ts
if (
    !botCache.helpers.isAdmin(message, targetGuildSettings)
) {
    return addReaction(message.channelID, message.id, "❌");
}
```

**Workaround**

Move the if statement header body up to the first line and reformat the file to get the expected output.

```ts
if (!botCache.helpers.isAdmin(message, targetGuildSettings)
) {
    return addReaction(message.channelID, message.id, "❌");
}
```

Reported by @skillz4killz

**Comment by kdy1** [comment]
I'll work on this.

**Comment by dsherret** [comment]
@kdy1 thanks!

**Comment by kdy1** [comment]
@dsherret Is it okay to create `StartInfo` of `IfStmt.cons` from `parse_if_stmt` and pass it down to the other methods like `parse_header_with_conditional_brace_body` and `parse_conditional_brace_body`?

I think there's a better way, but I'm asking this because the first one I can think of now is using the condition in `parse_if_stmt`  with the start info of `cons`.

**Comment by dsherret** [comment]
@kdy1 I'd have to spend more time looking at it (I was working all day today), but I believe that should be ok! Alternatively, perhaps storing in `context.store_info_for_node(node, info)` and `context.get_info_for_node(node)` would be useful, but maybe not.

**Comment by kdy1** [comment]
@dsherret I have some questions.

For first, Is there a way to preserve comments while parsing paren items twice? I tried parsing paren items twice with different options and comments are gone. 

Code: https://github.com/kdy1/dprint-plugin-typescript/blob/74b3a663cbfcbe35c83e27cb4dfac17d4894b507/src/parsing/parser.rs#L3492-L3512



The second question is, why does the newline come between `botCache` and `.helpers` in the test below?
I think `ifAboveWidth` should be false, but it seems to be true. Does the condition calculate width based on the input?

If then, is there a way to calculate lines based on output?



```
Condition: indentIfNotStartOfLine
  true:
    `(`
    Signal::StartNewLineGroup
    `!`
    Info: topMemberStart
    `botCache`
    Condition: ifAboveWidth # This line
      true:
        Signal::PossibleNewLine
    Info: topMemberEnd
    Condition: indentIfStartOfLine
      true:
        Signal::StartIndent
        `.`
        `helpers`
        Signal::FinishIndent
      false:
        `.`
        `helpers`
    Signal::FinishNewLineGroup
    Info: beforeTrailingComments
    Condition: newLineIfHasCommentsAndNotStartOfNewLine
      true:
        Signal::NewLine
    `)`
  false:
    `(`
    Info: surroundWithNewLinesIndentedIfMultiLineStart
    Condition: newlineIfMultiLine
      true:
        Signal::NewLine
        Signal::StartIndent
        `!`
        Info: topMemberStart
        `botCache`
        Condition: ifAboveWidth  # And this line
          true:
            Signal::PossibleNewLine
        Info: topMemberEnd
        Condition: indentIfStartOfLine
          true:
            Signal::StartIndent
            `.`
            `helpers`
            Signal::FinishIndent
          false:
            `.`
            `helpers`
        Signal::FinishIndent
        Signal::NewLine
      false:
        Condition: ifAboveWidth
          true:
            Signal::PossibleNewLine
        `!`
        Info: topMemberStart
        `botCache`
        Condition: ifAboveWidth
          true:
            Signal::PossibleNewLine
        Info: topMemberEnd
        Condition: indentIfStartOfLine
          true:
            Signal::StartIndent
            `.`
            `helpers`
            Signal::FinishIndent
          false:
            `.`
            `helpers`
    Info: surroundWithNewLineIndentedsIfMultiLineEnd
    Info: beforeTrailingComments
    Condition: newLineIfHasCommentsAndNotStartOfNewLine
      true:
        Signal::NewLine
    `)`

---
Failed:   should keep the header on one line if possible (only) (./tests/specs/statements/ifStatement/IfStatement_All.txt)
Expected: `"if (!botCache.helpers) {\n    return addReaction(message.channelID, message.id, \"❌\");\n}\n"`,
Actual:   `"if (!botCache\n    .helpers)\n{\n    return addReaction(\n        message.channelID,\n        message.id,\n        \"❌\",\n    );\n}\n"`,`
---

```






Btw, it is much harder than I expected.

**Comment by dsherret** [comment]
@kdy1 yeah, a lot of these problems are really hard :( This is probably the hardest project I've ever worked on and there still are a few areas in the design from dprint_core that I want to spend thinking about and refining. I'm not convinced the current approach used in from dprint_core will work in all scenarios and so probably a more complicated hybrid print item w/ tree structure would work best, but for now the design is ok.

Also, sorry for my delay.. I usually have to spend a bit of time looking into it before responding.

> For first, Is there a way to preserve comments while parsing paren items twice? I tried parsing paren items twice with different options and comments are gone.

No, right now the parser keeps track of all comments its parsed and only allows them to be parsed once (it prevents a comment being duplicated). I think here probably in order to support this, instead of adding `use_new_line_group` to `ParseNodeInParensOptions`, you could instead just wrap `test_items_different_line` with `with_new_line_group(test_items_different_line)`. Maybe that wouldn't work though. I'm not sure why it needs to be used in one scenario and not the other. Another way is instead of passing a boolean in this case, you could pass a condition. There is a `BoolOrCondition` in `dprint_core::formatting`.

> The second question is, why does the newline come between botCache and .helpers in the test below?
I think ifAboveWidth should be false, but it seems to be true. Does the condition calculate width based on the input?

I'm not sure exactly why that happens, but I can take a look at it tomorrow. Most likely that possible newline has the highest precedence somehow.

The `if_above_width` is to see if the current column is above the indentation width (indentation width is the width of an indent... ex. 2 spaces, 4 spaces, etc) and if so, make it a possible newline. When false, it prevents a possible newline from happening in scenarios like `t.testing` where it wouldn't make sense to have a newline where it would go to `t\n    .testing` if the current column number at the period is below or at the indentation width.

> If then, is there a way to calculate lines based on output?

The only way to check the output is to look at a resolved info in a condition's resolver (ex. `condition_context.get_resolved_info(&start_info)`) or to look at `condition_context.writer_info`. You can see the `if_above_width` implementation here and how it checks `condition_context.writer_info` to see the column number of where that info ended up:

https://github.com/dprint/dprint/blob/8e15c9a06386366fcb448199ca021b9536abf815/crates/core/src/formatting/conditions.rs#L96

Let me know if you would like more clarification 🙂

**Comment by kdy1** [comment]
Sorry for the late response. I was busy because my bundler had too many bugs.
And thank you for such a detailed response!

I think my questions are all resolved. I'll give it an another try.

**Comment by kdy1** [comment]
If I'm delaying the task, please feel free to take this over.

**Comment by kdy1** [comment]
Sorry for the delay.
I have another question.

I tried the reevaluation based approach, just like

https://github.com/kdy1/dprint-plugin-typescript/blob/deb5ed3210296a285a52b55d8b1962c4621d7bad/src/parsing/parser.rs#L5761-L5773

I thought it is about jumping back to the start of a parenthesized node. But it resulted in the same error.

My question is, what does `resolved` mean in `force_reevaluation_once_resolved`?

**Comment by dsherret** [comment]
@kdy1 sorry, I missed this. The `force_reevaluation_once_resolved` is a bit of a hack that abuses how conditions work:

```rs
/// This condition can be used to force the printer to jump back to the point
/// this condition exists at once the provided info is resolved.
pub fn force_reevaluation_once_resolved(info: Info) -> Condition {
    Condition::new("forceReevaluationOnceInfoResolved", ConditionProperties {
        condition: Rc::new(Box::new(move |context| {
            let resolved_info = context.get_resolved_info(&info);
            if resolved_info.is_some() { Some(false) } else { None }
        })),
        true_path: None,
        false_path: None,
    })
}
```

When a condition cannot be resolved due to an info not being resolved, the printer will say "I can't resolve this condition yet because this info hasn't been resolved yet. Remember this condition and jump back to it when this info is resolved.". Then once it encounters the info specified, it will jump back to printing where the condition is located. So, in reality, this condition is not being used as a condition... I need to spend some time thinking more about the underlying print items to make scenarios like this more clear.

Also, I just looked into this issue. The main fix is to make the header portion be in a newline group and then ensure the space after the header is what causes a newline when the single-line body exceeds the line width. It was a bit of a mess in this area of the code and it wasn't properly writing a space after the header to cause the newline to occur. I went ahead and did a fix in #118.
